packages/vite/src/node/build.ts

```diff
@@ -1247,3 +1367,137 @@ function areSeparateFolders(a: string, b: string) {
     !nb.startsWith(withTrailingSlash(na))
   )
 }
+
+export class BuildEnvironment extends Environment {
+  mode = 'build' as const
+
+  constructor(
+    name: string,
+    config: ResolvedConfig,
+    setup?: {
+      options?: EnvironmentOptions
+    },
+  ) {
+    // TODO: move this to the base Environment class?
+    let options =
+      config.environments[name] ?? getDefaultResolvedEnvironmentOptions(config)
+    if (setup?.options) {
+      options = mergeConfig(
+        options,
+        setup?.options,
+      ) as ResolvedEnvironmentOptions
+    }
+    super(name, config, options)
+  }
+}
+
+export interface ViteBuilder {
+  environments: Record<string, BuildEnvironment>
+  config: ResolvedConfig
+  buildEnvironments(): Promise<void>
+  build(
+    environment: BuildEnvironment,
+  ): Promise<RollupOutput | RollupOutput[] | RollupWatcher>
+}
+
+export interface BuilderOptions {
+  buildEnvironments?: (
+    builder: ViteBuilder,
+    build: (environment: BuildEnvironment) => Promise<void>,
+  ) => Promise<void>
+}
```

Comment:
Is the 2nd argument of `BuilderOptions.buildEnvironments` necessary?
I currently used it as

https://github.com/vitejs/vite/blob/4468c79d6f707757d0557f7e0dd432cf54cb38ef/playground/environment-react-ssr/vite.config.ts#L49-L54

but it looks like this is equivalent to using 1st argument's `builder.bulid`:

```ts
   async buildEnvironments(builder) { 
     await builder.build(builder.environments.client) 
     await builder.build(builder.environments.ssr) 
   }, 
```

The difference is that actually `builder.build` returns `RollupOutput` but the 2nd argument version return `void`, so it makes less useful.

Reply:
Yes. Maybe we should rename it to `builder: { buildApp(builder) }`. So we are more clear that you may want to also do other things in this hook apart from building the environments.

There are also some discussions about deprecating `build.lib` in favor of a new `libraries` top level entry to let user define any number of app libraries (each with a different environment options config). So `build --app` may also end up building all libraries too in the future.

Reply:
Thanks, that looks good to me.

I don't have a particular preference on naming `buildEnvironments` and `--all` vs. `buildApp` and `--app`.